tools/bin/mbedtls-run-tests

```diff
@@ -0,0 +1,159 @@
+        return (re.match(self.include, description) and
+                not re.match(self.exclude, description))
+
+
+def extract_description(stanza):
+    """Extract the description from a .data stanza."""
+    m = re.match(r'(?:\n|#[^\n]*\n)*([^#\n][^\n]*)\n', stanza + '\n')
+    if m:
+        return m.group(1)
+    else:
+        return None
+
+def filter_test_cases(all_datax, tcf, temp_datax):
+    """Filter test cases.
+
+    Filter test cases from datax_file based on their description according
+    to the filter tcf. Write the selected test cases to temp_datax.
+    """
+    in_stanza = False
+    in_match = False
+    for line in open(all_datax, 'rb'):
+        was_in_stanza = in_stanza
+        in_stanza = not not line.strip()
+        if in_stanza and not was_in_stanza:
+            in_match = tcf.match(line)
+        elif not in_stanza:
+            in_match = False
+        if in_match:
+            temp_datax.write(line)
+    temp_datax.flush()
+
+def run_exe(keep_temp, precommand, exe, extra_options, all_datax, tcf):
+    """Run one Mbed TLS test suites based on the specified options.
+
+    See `main` for what options are valid.
+    """
+    directory = os.path.dirname(exe)
+    with tempfile.NamedTemporaryFile(
+            dir=directory, suffix='.datax', delete=not keep_temp
+    ) as temp_datax:
+        filter_test_cases(all_datax, tcf, temp_datax)
+        print(directory,
+              os.path.basename(exe),
+              os.path.basename(temp_datax.name))
+        cmd = (precommand +
+               [os.path.join(os.path.curdir, os.path.basename(exe))] +
+               extra_options +
+               [os.path.basename(temp_datax.name)])
+        outcome = subprocess.run(cmd, cwd=directory)
+    return outcome.returncode
+
+def find_data_file(exe):
+    """Return the .datax file for the specified test suite executable."""
+    directory = os.path.dirname(exe)
+    basename = os.path.basename(exe)
+    basebasename = os.path.splitext(basename)[0]
+    for datax_file in [
+            os.path.join(directory, basename + '.datax'),
+            os.path.join(directory, basebasename + '.datax'),
+    ]:
+        if os.path.exists(datax_file):
+            return datax_file
+    else:
+        raise Exception('.datax file not found for ' + exe)
+
+def run(options):
+    """Run Mbed TLS test suites based on the specified options.
+
+    See `main` for what options are valid.
+    """
+    extra_options = []
+    if options.verbose:
+        extra_options.append('-v')
+    tcf = TestCaseFilter(options)
+    # Make sure we can find all the data files before we start running tests.
+    data_files = [find_data_file(exe) for exe in options.exes]
+    global_status = 0
+    for exe, data_file in zip(options.exes, data_files):
+        status = run_exe(options.keep,
+                         options.command, exe, extra_options,
+                         data_file, tcf)
+        if status > global_status:
```

Comment:
Negative return codes (meaning the process was terminated by a signal) are possible - should probably catch this too (e.g. status = abs(run_exe ...)